### DISCUSSION TITLE
PC to Hub comminication

### DISCUSSION THREAD

**Discussion by jeagh** [pybricks/support#997] [Q&A]
I had this sorted until I upgraded the Robot Inventor hub firmware now I get this error:

'NoneType' object has no attribute 'lower'

I had modified the example so that I could send degrees for motor rotation and so I thought I'd made an error somewhere, however, I now get the same error message with the original example. Has the firmware upgrade caused this error and if so is there a workaround or another way to communicate with the hub? I don't need Bluetooth, a USB connection is OK.

**Comment by dlech** [comment]
Can you share the code the reproduce the problem and give the full stack track of the error?

**Reply by jeagh** [reply]
Thank you for your reply. The problem was that I didn't disconnect from the Pybricks IDE and I sometimes missed the "press button now" command. Still, I don't have this working quite correctly. The following is a slightly modified version of the example code. The laptop Python code follows.
```python
# NOTE: Run this program with the latest
# firmware provided via https://beta.pybricks.com/

from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll

motor = Motor(Port.D)

# Optional: Register stdin for polling. This allows
# you to wait for incoming data without blocking.
keyboard = poll()
keyboard.register(stdin)

while True:

    # Optional: Check available input.
    while not keyboard.poll(0):
        # Optional: Do something here.
        wait(10)

    # Read three bytes.
    cmd = stdin.buffer.read(3)
    print(cmd)
    wait(500)

    degree = int(cmd)
    print(degree)
    motor.track_target(degree)
    wait(1000)

     # Send a response.
    stdout.buffer.write(b"OK")

```

```python
        # Snippet from laptop code

        # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")

        # Send a few messages to the hub.
        #await send(client, b"fwd")
        await send(client, b'000')
        await asyncio.sleep(2)
        
        await send(client, b'090')
        await asyncio.sleep(2)
        
        await send(client, b'180')
        await asyncio.sleep(2)
        
        await send(client, b'000')
        await asyncio.sleep(2)
   
```
I don't really want the sleep and wait delays, they're just added to try to at least get something working. What would be better is to have the hub send OK to the laptop when it's ready for the next degree command.

Finally, the motor does not stop between commands, instead it rocks back and forth slightly. Can this be prevented or is it something that inherit in the motor design?

**Reply by dlech** [reply]
> Finally, the motor does not stop between commands, instead it rocks back and forth slightly. Can this be prevented or is it something that inherit in the motor design?

Which firmware version are you using? There has been some improvement to this recently.

**Reply by jeagh** [reply]
I upgraded the hub firmware last week. How do I tell what version I have?

**Reply by dlech** [reply]
```python
from pybricks import version
print(version)
```

**Reply by jeagh** [reply]
('primehub', '3.2.3', 'v3.2.3 on 2023-02-17')

**Reply by dlech** [reply]
Try the v3.3 beta firmware from https://beta.pybricks.com.

**Reply by jeagh** [reply]
OK

**Reply by jeagh** [reply]
The rocking seems to have been eliminated or much reduced. Thank you for that.

I've had some more thought on my original question but before I continue, possibly in the wrong direction, do you have the time to look over the code that I posted?

**Reply by dlech** [reply]
It is still not clear to me where the error is coming from without seeing the stack trace.

**Reply by jeagh** [reply]
Sorry for my clumsiness.  I've eliminated the error message, I'm now wondering how I might improve the code that I posted earlier.

**Reply by dlech** [reply]
```python
# SPDX-License-Identifier: MIT
# Copyright (c) 2020 Henrik Blidh
# Copyright (c) 2022 The Pybricks Authors

import asyncio
from bleak import BleakScanner, BleakClient

UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"

# Replace this with the name of your hub if you changed
# it when installing the Pybricks firmware.
HUB_NAME = "Pybricks Hub"


async def main():
    def hub_filter(device, ad):
        return device.name and device.name.lower() == HUB_NAME.lower()

    def handle_disconnect(_):
        print("Hub was disconnected.")

    ok_event = asyncio.Event()

    def handle_rx(_, data: bytearray):
        if data == b"OK":
            ok_event.set()

    # Find the device and initialize client.
    device = await BleakScanner.find_device_by_filter(hub_filter)
    
    async with BleakClient(device, disconnected_callback=handle_disconnect) as client:

        # Shorthand for sending some data to the hub.
        async def send(data):
            await client.write_gatt_char(UART_RX_CHAR_UUID, data)

        # Subscribe to notifications
        await client.start_notify(UART_TX_CHAR_UUID, handle_rx)

        # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")

        # Send a few messages to the hub.
        ok_event.clear()
        await send(b'000')
        await ok_event.wait()

        ok_event.clear()
        await send(b'090')
        await ok_event.wait()
        
        ok_event.clear()
        await send(b'180')
        await ok_event.wait()
        
        ok_event.clear()
        await send(b'000')
        await ok_event.wait()


# Run the main async program.
asyncio.run(main())

```

**Reply by jeagh** [reply]
Thank you yet again, I should be able to make some progress now.

The laptop extracts the x and y values from a gcode file and, at the moment, I'm working on sending just the x values to the hub. The plan is to send the x value, wait for OK, and then send the y value. Repeat until end of file.
